### ISSUE TITLE
"OSError: Failed to read from device" when accessing microphone from non-root user on Raspberry Pi

### ISSUE THREAD

**Issue by Darkhand81** [Picovoice/pvrecorder#82]
Hello!

First off, love the project!

Hardware: Raspberry Pi 4 - 2GB
Software: Pi OS 64-bit, Debian Bookworm base, 061223 image, fully updated as of 8-25-23, default install.
Microphone: Realtek USB Microphone - USB ID 0bda:4938

I'm not entirely sure if this is an issue with the Pi's default permissions, or the way pvrecorder accesses the microphone on the Pi, as the microphone does work in other instances.

I've confirmed the correct audio device is being accessed:

```
pi@raspberrypi:~ $ porcupine_demo_mic --show_audio_devices 
Device 0: Monitor of Built-in Audio Analog Stereo
Device 1: Monitor of Built-in Audio Digital Stereo (HDMI)
Device 2: Realtek Audio USB Analog Stereo  <-- (This device)
pi@haus-ear-01:~ $ 
```

But attempting to run pvrecorder (in any instance, but specifically the porcupine demo for this example) results in the Microphone not being able to be accessed:

```
pi@raspberrypi:~ $ porcupine_demo_mic --access_key {MYKEYHERE} --keywords porcupine --audio_device_index 2
Porcupine version: 2.2.0
Listening ... (press Ctrl+C to exit)
Traceback (most recent call last):
  File "/home/pi/.local/bin/porcupine_demo_mic", line 8, in <module>
    sys.exit(main())
  File "/home/pi/.local/lib/python3.9/site-packages/pvporcupinedemo/porcupine_demo_mic.py", line 140, in main
    pcm = recorder.read()
  File "/home/pi/.local/lib/python3.9/site-packages/pvrecorder/_pvrecorder.py", line 178, in read
    raise self._PVRECORDER_STATUS_TO_EXCEPTION[status]("Failed to read from device.")
OSError: Failed to read from device.
```

However, I've also confirmed the device does work when manually recording:

```
pi@raspberrypi:~ $ arecord --format=S16_LE --duration=5 --rate=16000 --file-type=wav test.wav
```

The resulting file is valid and audio is recorded correctly.

Interestingly, creating a python venv under root and installing everything there _does_ work, initially leading me to think it was a simple permissions issue:

```
pi@raspberrypi:/dev/snd $ ll
total 0
drwxr-xr-x  2 root root        60 Aug 25 17:41 by-id
drwxr-xr-x  2 root root       120 Aug 25 17:41 by-path
crw-rw----+ 1 root audio 116,   0 Aug 25 16:16 controlC0
crw-rw----+ 1 root audio 116,  32 Aug 25 17:41 controlC1
crw-rw----+ 1 root audio 116,  64 Aug 25 16:16 controlC2
crw-rw----+ 1 root audio 116,  96 Aug 25 16:16 controlC3
crw-rw----+ 1 root audio 116,  16 Aug 25 16:16 pcmC0D0p
crw-rw----+ 1 root audio 116,  56 Aug 25 17:41 pcmC1D0c
crw-rw----+ 1 root audio 116,  80 Aug 25 16:16 pcmC2D0p
crw-rw----+ 1 root audio 116, 112 Aug 25 16:16 pcmC3D0p
crw-rw----+ 1 root audio 116,   1 Aug 25 16:16 seq
crw-rw----+ 1 root audio 116,  33 Aug 25 16:16 timer
pi@raspberrypi:/dev/snd $ 
```

All sound devices are owned by `root`, and accessible to the `audio` group.

The `pi` user is a member of this group, however:

```
pi@haus-ear-01:~ $ groups pi
pi : pi adm dialout cdrom sudo audio video plugdev games users input render netdev spi i2c gpio lpadmin
pi@haus-ear-01:~ $ 
```

Attempting to manually record from the raw audio device in question results in an error as well, but interestingly this same error occurs when manually recording under root, even though pvrecorder did work properly there:

```
pi@raspberrypi:/dev/snd $ dd if=/dev/snd/pcmC1D0c of=~/audio.raw
dd: error reading '/dev/snd/pcmC1D0c': File descriptor in bad state
```

The issue seems to have popped up with others lately too, like [this](https://forums.raspberrypi.com/viewtopic.php?t=354626) issue I found on the Pi forums.

The question I haven't been able to answer, is how does `pvrecorder` access the microphone differently than the `arecord` application?  That method can access the device just fine under the `pi` user, while `pvrecorder` can't.

I'm leaning toward this being a default configuration issue with the current Pi OS vs a pvrecorder bug, but as the Pi is a common way to use the software, I thought you might be able to provide some insight.  Let me know if there's any other info or testing I can provide.

Thanks in advance!

**Comment by ksyeo1010** [comment]
Hi, can you try again with the latest demo package for python? (`pvporcupinedemo==2.2.3`)

**Comment by Darkhand81** [comment]
> Hi, can you try again with the latest demo package for python? (`pvporcupinedemo==2.2.3`)

This did fix it! 2.2.3 runs perfectly on the Pi4.  Thanks so much!
